ssl/statem/statem_clnt.c

```diff
@@ -1787,12 +1787,24 @@ MSG_PROCESS_RETURN tls_process_server_hello(SSL_CONNECTION *s, PACKET *pkt)
      * we're done with this message
      */
     if (SSL_CONNECTION_IS_TLS13(s)) {
-        if (!ssl->method->ssl3_enc->setup_key_block(s)
-                || !ssl->method->ssl3_enc->change_cipher_state(s,
-                    SSL3_CC_HANDSHAKE | SSL3_CHANGE_CIPHER_CLIENT_READ)) {
+        if (!ssl->method->ssl3_enc->setup_key_block(s))
             /* SSLfatal() already called */
             goto err;
-        }
+        if (!ssl->method->ssl3_enc->change_cipher_state(s,
+             SSL3_CC_HANDSHAKE | SSL3_CHANGE_CIPHER_CLIENT_READ))
+            /* SSLfatal() already called */
+            goto err;
+        /*
+         * QUIC needs to provide the write handshake at the same time as the read handshake
+         * in case the quic stack needs to ACK packets independent of the handshake process
+         */
+        if (SSL_IS_QUIC_HANDSHAKE(s)
+            && s->early_data_state != SSL_EARLY_DATA_NONE
+            && !ssl->method->ssl3_enc->change_cipher_state(s,
+             SSL3_CC_HANDSHAKE | SSL3_CHANGE_CIPHER_CLIENT_WRITE))
+            /* SSLfatal() already called */
+            goto err;
```

Comment:
I'm fine with this change in principle. However, probably there is more work to do here. We also yield this secret in these other locations:

https://github.com/openssl/openssl/blob/b9282ecafa36bf3192bdc2cd54c729fb70a92814/ssl/statem/statem_clnt.c#L3813-L3830

https://github.com/openssl/openssl/blob/b9282ecafa36bf3192bdc2cd54c729fb70a92814/ssl/statem/statem_clnt.c#L3903-L3919

https://github.com/openssl/openssl/blob/00480f1def86aab882f2bd4d15d2d03862009718/ssl/statem/statem_lib.c#L621-L636

We must only yield the secret once. So probably the other locations need to be adjusted accordingly.

Reply:
I think I've fixed those up now